consumers/campaigner/main.go

```diff
@@ -44,150 +47,112 @@ func (h *MessageHandler) HandleMessage(m *nsq.Message) error {
 		return nil
 	}
 
-	// fetching subs that are active and that have not been blacklisted
+	campaign, err := h.s.GetCampaign(msg.CampaignID, msg.UserID)
+	if err != nil {
+		logrus.WithFields(logrus.Fields{
+			"campaign_id": msg.CampaignID,
+			"user_id":     msg.UserID},
+		).WithError(err).Error("unable to find campaign")
+		return nil
+	}
+	if campaign.Status != entities.StatusDraft {
+		return nil
+	}
+
+	campaign.Status = entities.StatusSending
+	err = h.s.UpdateCampaign(campaign)
+	if err != nil {
+		return err
+	}
+
 	var (
-		active      = true
-		blacklisted = false
-		timestamp   time.Time
-		nextID      int64
-		limit       int64 = 1000
+		timestamp time.Time
+		nextID    int64
+		limit     int64 = 1000
 	)
-	for {
-		subs, err := h.s.GetDistinctSubscribersBySegmentIDs(
-			msg.SegmentIDs,
-			msg.UserID,
-			blacklisted,
-			active,
-			timestamp,
-			nextID,
-			limit,
-		)
+
+	subs, err := h.s.GetDistinctSubscribersBySegmentIDs(
+		msg.SegmentIDs,
+		msg.UserID,
+		false,
+		true,
+		timestamp,
+		nextID,
+		limit,
+	)
+	if err != nil {
+		logrus.WithFields(logrus.Fields{
+			"user_id":     msg.UserID,
+			"segment_ids": msg.SegmentIDs,
+		}).WithError(err).Error("Unable to fetch subscribers.")
+		return nil
+	}
+
+	if len(subs) == 0 {
+		return nil
+	}
+
+	template, err := h.svc.GetTemplate(context.Background(), msg.TemplateID, msg.UserID)
+	if err != nil {
+		logrus.WithError(err).
+			WithField("template_id", msg.TemplateID).
+			Error("unable to get template")
+		return nil
+	}
+	html, _ := mustache.ParseString(template.HTMLPart)
+	txt, _ := mustache.ParseString(template.SubjectPart)
+	sub, _ := mustache.ParseString(template.TextPart)
+
+	for _, s := range subs {
+		m, err := s.GetMetadata()
 		if err != nil {
-			logrus.WithFields(logrus.Fields{
-				"user_id":     msg.UserID,
-				"segment_ids": msg.SegmentIDs,
-			}).WithError(err).Error("Unable to fetch subscribers.")
-			break
+			logrus.WithError(err).
+				WithField("subscriber", s).
+				Error("Unable to get subscriber metadata.")
+			continue
 		}
-
-		if len(subs) == 0 {
-			break
+		//  todo merge def template data with sub metadata
+
+		// fill template buffer with rendered template.
+		var buf bytes.Buffer
```

Comment:
We'll need separate buffers for html, text and subject. Also we should send them separately in the SenderConsumerParams.